lib/cocina/models/builders/title_builder.rb

```diff
@@ -27,92 +27,127 @@ def self.build(titles, strategy: :first, add_punctuation: true)
           cocina_title = other_title(titles) if cocina_title.blank?
 
-          cocina_title = cocina_title.first if cocina_title.is_a?(Array)
           extract_main_title(cocina_title)
         end
 
         private
 
         attr_reader :strategy
 
+        # rubocop:disable Metrics/CyclomaticComplexity
+        # rubocop:disable Metrics/MethodLength
+        # rubocop:disable Metrics/PerceivedComplexity
         def extract_title(cocina_title)
+          title_values = if cocina_title.value
+                           cocina_title.value
+                         elsif cocina_title.structuredValue.present?
+                           rebuild_structured_value(cocina_title)
+                         elsif cocina_title.parallelValue.present?
+                           primary = cocina_title.parallelValue.find { |pvalue| pvalue.status == 'primary' }
+                           if primary && only_one_parallel_value? && strategy == :first
+                             extract_title(primary)
+                           elsif only_one_parallel_value? && strategy == :first
+                             untyped = cocina_title.parallelValue.find { |pvalue| pvalue.type.blank? }
+                             extract_title(untyped || cocina_title.parallelValue.first)
+                           else
+                             cocina_title.parallelValue.map { |pvalue| extract_title(pvalue) }
+                           end
+                         end
+          result = [title_values].flatten.compact.map { |val| remove_trailing_punctuation(val.strip) }
+          result.length == 1 ? result.first : result
+        end
+        # rubocop:enable Metrics/CyclomaticComplexity
+        # rubocop:enable Metrics/MethodLength
+        # rubocop:enable Metrics/PerceivedComplexity
+
+        # @return [Array<String>] the main title value(s) for Solr - can be array due to parallel titles
+        def extract_main_title(cocina_title) # rubocop:disable Metrics/PerceivedComplexity
           result = if cocina_title.value
-                     cocina_title.value
+                     cocina_title.value # covers both title and main title types
                    elsif cocina_title.structuredValue.present?
-                     title_from_structured_values(cocina_title)
+                     main_title_from_structured_values(cocina_title)
                    elsif cocina_title.parallelValue.present?
-                     return build(cocina_title.parallelValue)
+                     primary = cocina_title.parallelValue.find { |pvalue| pvalue.status == 'primary' }
+                     if primary
+                       extract_main_title(primary)
+                     else
+                       cocina_title.parallelValue.map { |pvalue| extract_main_title(pvalue) }
+                     end
                    end
-          remove_trailing_punctuation(result.strip) if result.present?
-        end
+          return [] if result.blank?
```

Comment:
this is one of my "new" methods (though you can see it was already added to the code base) and it is less convoluted because it isn't trying to keep old behavior the same.  Sigh.